# Turicreate baseline
Более подробно [почитать](https://apple.github.io/turicreate/docs/userguide/recommender/) про библиотеку

In [1]:
import os
import pandas as pd
import numpy as np
import turicreate as tc

In [2]:
# precision at k
def precision_at_k(r_true_arr, k):
    return np.sum(r_true_arr[:k]) / k


# average precision at k
def average_precision_at_k(r_true_arr, k):
    apk = 0
    for n in range(0, k):
        apk += precision_at_k(r_true_arr, n + 1) * r_true_arr[n]
    if np.sum(r_true_arr[:k]) != 0:
        return (apk) / k
    else:
        return 0


# average normed precision at k
def average_normed_precision_at_k(r_true_arr, k, n_true):
    apk = 0
    apk_ideal = n_true / k
    
    for n in range(0, k):
        apk += precision_at_k(r_true_arr, n + 1) * r_true_arr[n]
    if np.sum(r_true_arr[:k]) != 0:
        return ((apk) / k) / apk_ideal
    else:
        return 0

In [3]:
df_clients = tc.SFrame.read_csv('../retailhero-uplift/data/clients.csv')
df_products = tc.SFrame.read_csv("../retailhero-uplift/data/products.csv")
df_purchases = tc.SFrame.read_csv("../retailhero-uplift/data/purchases.csv")

Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/clients.csv

Parsing completed. Parsed 100 lines in 0.318353 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str,str,int,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/clients.csv

Parsing completed. Parsed 400162 lines in 0.367728 secs.

Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/products.csv

Parsing completed. Parsed 100 lines in 0.064703 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str,str,str,str,float,str,str,float,int,int]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/products.csv

Parsing completed. Parsed 43038 lines in 0.06195 secs.

Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/purchases.csv

Parsing completed. Parsed 100 lines in 0.5527 secs.

------------------------------------------------------
Inferred types from first 100 line(s) of file as 
column_type_hints=[str,str,str,float,float,float,float,float,str,str,float,float,str]
If parsing fails due to incorrect types, you can correct
the inferred type list above and pass it to read_csv in
the column_type_hints argument
------------------------------------------------------


Read 537642 lines. Lines per second: 598212

Read 4840156 lines. Lines per second: 786821

Read 9142422 lines. Lines per second: 777629

Read 13444938 lines. Lines per second: 781275

Read 17209972 lines. Lines per second: 770580

Read 20974153 lines. Lines per second: 765850

Read 24738333 lines. Lines per second: 760468

Read 29040839 lines. Lines per second: 761094

Read 33343168 lines. Lines per second: 764752

Read 37107078 lines. Lines per second: 762250

Read 40871601 lines. Lines per second: 754206

Read 45173555 lines. Lines per second: 758605

Finished parsing file /Users/arty.erokhin/projects/competition/retail_hero/retailhero-uplift/data/purchases.csv

Parsing completed. Parsed 45786568 lines in 60.1803 secs.

In [4]:
df_purchases[:2]

Columns:
	client_id	str
	transaction_id	str
	transaction_datetime	str
	regular_points_received	float
	express_points_received	float
	regular_points_spent	float
	express_points_spent	float
	purchase_sum	float
	store_id	str
	product_id	str
	product_quantity	float
	trn_sum_from_iss	float
	trn_sum_from_red	str

Rows: 2

Data:
+------------+----------------+----------------------+-------------------------+
| client_id  | transaction_id | transaction_datetime | regular_points_received |
+------------+----------------+----------------------+-------------------------+
| 000012768d |   7e3e2e3984   | 2018-12-01 07:12:45  |           10.0          |
| 000012768d |   7e3e2e3984   | 2018-12-01 07:12:45  |           10.0          |
+------------+----------------+----------------------+-------------------------+
+-------------------------+----------------------+----------------------+--------------+
| express_points_received | regular_points_spent | express_points_spent | purchase_sum |
+-------------------------+----------------------+----------------------+--------------+
|           0.0           |         0.0          |         0.0          |    1007.0    |
|           0.0           |         0.0          |         0.0          |    1007.0    |
+-------------------------+----------------------+----------------------+--------------+
+------------+------------+------------------+------------------+------------------+
|  store_id  | product_id | product_quantity | trn_sum_from_iss | trn_sum_from_red |
+------------+------------+------------------+------------------+------------------+
| 54a4a11a29 | 9a80204f78 |       2.0        |       80.0       |                  |
| 54a4a11a29 | da89ebd374 |       1.0        |       65.0       |                  |
+------------+------------+------------------+------------------+------------------+
[2 rows x 13 columns]

In [5]:
df_clients[:2]

Columns:
	client_id	str
	first_issue_date	str
	first_redeem_date	str
	age	int
	gender	str

Rows: 2

Data:
+------------+---------------------+---------------------+-----+--------+
| client_id  |   first_issue_date  |  first_redeem_date  | age | gender |
+------------+---------------------+---------------------+-----+--------+
| 000012768d | 2017-08-05 15:40:48 | 2018-01-04 19:30:07 |  45 |   U    |
| 000036f903 | 2017-04-10 13:54:23 | 2017-04-23 12:37:56 |  72 |   F    |
+------------+---------------------+---------------------+-----+--------+
[2 rows x 5 columns]

In [6]:
df_products[:2]

Columns:
	product_id	str
	level_1	str
	level_2	str
	level_3	str
	level_4	str
	segment_id	float
	brand_id	str
	vendor_id	str
	netto	float
	is_own_trademark	int
	is_alcohol	int

Rows: 2

Data:
+------------+------------+------------+------------+------------+------------+
| product_id |  level_1   |  level_2   |  level_3   |  level_4   | segment_id |
+------------+------------+------------+------------+------------+------------+
| 0003020d3c | c3d3a8e8c6 | c2a3ea8d5e | b7cda0ec0c | 6376f2a852 |   123.0    |
| 0003870676 | e344ab2e71 | 52f13dac0c | d3cfe81323 | 6dc544533f |   105.0    |
+------------+------------+------------+------------+------------+------------+
+------------+------------+-------+------------------+------------+
|  brand_id  | vendor_id  | netto | is_own_trademark | is_alcohol |
+------------+------------+-------+------------------+------------+
| 394a54a7c1 | 9eaff48661 |  0.4  |        0         |     0      |
| acd3dd483f | 10486c3cf0 |  0.68 |        0         |     0      |
+------------+------------+-------+------------------+------------+
[2 rows x 11 columns]

In [7]:
# сделаем значения текстовыми, чтобы потом Turicreate сам сделал OHE
df_products["segment_id"] = df_products["segment_id"].astype(str)
df_products["is_alcohol"] = df_products["is_alcohol"].astype(str)
df_products["is_own_trademark"] = df_products["is_own_trademark"].astype(str)

In [8]:
# разбиваем на трейн и тест
training_data, validation_data = \
    tc.recommender.util.random_split_by_user(df_purchases, "client_id", "product_id", 
                                            max_num_users=10000, random_seed=42)

In [9]:
training_data.shape, validation_data.shape

((45556268, 13), (230300, 13))

In [10]:
columns = ['client_id',
'product_id']

item_columns = ['product_id',
 'level_1',
 'level_2',
 'level_3',
 'level_4',
 'segment_id',
 'brand_id',
 'vendor_id',
 'is_own_trademark',
 'is_alcohol']

In [11]:
# обучим модель с дополнительными данными о продуктах
model = tc.recommender.ranking_factorization_recommender.create(training_data[columns], 
                                                        user_id="client_id", 
                                                        item_id="product_id",
                                                        max_iterations=20,
                                                        item_data=df_products[item_columns])

Preparing data set.

Data has 45556268 observations with 400157 users and 43038 items.

Data prepared in: 15.5776s

Training ranking_factorization_recommender for recommendations.

+--------------------------------+--------------------------------------------------+----------+

| Parameter                      | Description                                      | Value    |

+--------------------------------+--------------------------------------------------+----------+

| num_factors                    | Factor Dimension                                 | 32       |

| regularization                 | L2 Regularization on Factors                     | 1e-09    |

| solver                         | Solver used for training                         | adagrad  |

| linear_regularization          | L2 Regularization on Linear Coefficients         | 1e-09    |

| binary_target                  | Assume Binary Targets                            | True     |

| side_data_factorization        | Assign Factors for Side Data                     | True     |

| max_iterations                 | Maximum Number of Iterations                     | 20       |

+--------------------------------+--------------------------------------------------+----------+

Optimizing model using SGD; tuning step size.

Using 5694533 / 45556268 points for tuning the step size.

+---------+-------------------+------------------------------------------+

| Attempt | Initial Step Size | Estimated Objective Value                |

+---------+-------------------+------------------------------------------+

| 0       | 4.54545           | Not Viable                               |

| 1       | 1.13636           | 1.11669                                  |

| 2       | 0.568182          | 0.757816                                 |

| 3       | 0.284091          | 0.608791                                 |

| 4       | 0.142045          | 0.594699                                 |

| 5       | 0.0710227         | 0.662951                                 |

| 6       | 0.0355114         | 0.841015                                 |

| 7       | 0.0177557         | 1.01282                                  |

+---------+-------------------+------------------------------------------+

| Final   | 0.142045          | 0.594699                                 |

+---------+-------------------+------------------------------------------+

Starting Optimization.

+---------+--------------+-------------------+-----------------------------------+-------------+

| Iter.   | Elapsed Time | Approx. Objective | Approx. Training Predictive Error | Step Size   |

+---------+--------------+-------------------+-----------------------------------+-------------+

| Initial | 429us        | 1.38669           | 0.693539                          |             |

+---------+--------------+-------------------+-----------------------------------+-------------+

| 1       | 1m 3s        | 0.757868          | 0.369047                          | 0.142045    |

| 2       | 2m 5s        | 0.69208           | 0.329348                          | 0.142045    |

| 3       | 3m 7s        | 0.652203          | 0.30841                           | 0.142045    |

| 4       | 4m 10s       | 0.631805          | 0.297769                          | 0.142045    |

| 5       | 5m 13s       | 0.619062          | 0.291034                          | 0.142045    |

| 7       | 7m 17s       | 0.60504           | 0.282531                          | 0.142045    |

| 8       | 8m 19s       | 0.600173          | 0.279528                          | 0.142045    |

| 9       | 9m 28s       | 0.596225          | 0.277102                          | 0.142045    |

| 10      | 10m 36s      | 0.59294           | 0.275079                          | 0.142045    |

| 11      | 11m 43s      | 0.590259          | 0.27338                           | 0.142045    |

| 12      | 12m 51s      | 0.587607          | 0.271774                          | 0.142045    |

| 13      | 13m 55s      | 0.585626          | 0.270483                          | 0.142045    |

| 14      | 15m 1s       | 0.58372           | 0.269275                          | 0.142045    |

| 15      | 16m 10s      | 0.582131          | 0.268255                          | 0.142045    |

| 16      | 17m 14s      | 0.580686          | 0.267323                          | 0.142045    |

| 17      | 18m 19s      | 0.579183          | 0.266364                          | 0.142045    |

| 18      | 19m 24s      | 0.577796          | 0.265546                          | 0.142045    |

| 19      | 20m 28s      | 0.576662          | 0.264814                          | 0.142045    |

| 20      | 21m 39s      | 0.575677          | 0.264154                          | 0.142045    |

+---------+--------------+-------------------+-----------------------------------+-------------+

Optimization Complete: Maximum number of passes through the data reached.

Computing final objective value and training Predictive Error.

Final objective value: 0.574467

Final training Predictive Error: 0.263977

In [12]:
# рекомендации для пользователей из вадидационной выборки
recommended = model.recommend(validation_data["client_id"].unique(), k=30, 
                              exclude_known=False)

recommendations finished on 1000/9871 queries. users per second: 295.074

recommendations finished on 2000/9871 queries. users per second: 292.381

recommendations finished on 3000/9871 queries. users per second: 291.811

recommendations finished on 4000/9871 queries. users per second: 292.018

recommendations finished on 5000/9871 queries. users per second: 293.52

recommendations finished on 6000/9871 queries. users per second: 293.936

recommendations finished on 7000/9871 queries. users per second: 296.749

recommendations finished on 8000/9871 queries. users per second: 299.506

recommendations finished on 9000/9871 queries. users per second: 300.654

In [13]:
# датафрейм с покупками в реальности
reality = validation_data["client_id", "product_id"].to_dataframe()
reality.loc[:, "is_buyed"] = 1

# Metrics (precision@30, avg_precision@30, average_normed_precision@30)

In [14]:
rec_df = recommended.to_dataframe().merge(reality, on=["client_id", "product_id"], 
                                          how="left", sort=False).fillna(0)

In [15]:
# словарь с количеством покупок на валидации
real_dict = reality.groupby(by="client_id")["is_buyed"].sum().to_dict()

In [16]:
np.mean([precision_at_k(i, 30) for i in 
         rec_df.groupby(by="client_id", sort=False)["is_buyed"].apply(list)])

0.15021443285043734

In [17]:
np.mean([average_precision_at_k(i, 30) for i in 
         rec_df.groupby(by="client_id")["is_buyed"].apply(list)])

0.08843787115556434

In [18]:
np.mean([average_normed_precision_at_k(i, 30, real_dict.get(client, 0)) for client, i in 
         rec_df.groupby(by="client_id")["is_buyed"].apply(list).reset_index().values])

0.09879876704365677

# Сохраняем нашу модель

In [19]:
model.save("x5.model")

# Загрузка модели быстрая

In [20]:
%%time
model = tc.load_model("x5.model")

CPU times: user 770 ms, sys: 149 ms, total: 919 ms
Wall time: 480 ms


# Итого:
1. Мы обучили factorization_recommender из библиотеки turicreate;
2. Сделали валидационное и обучающее множества, получив на валидации результаты:
    - average_precision_at_k: 0.088
    - average_normed_precision_at_k: 0.099
3. Сохранили модель и использовали ее в решении;
4. Получили следующие результаты при загрузке нашего ответа:
    - check: 0.0907
    - public: 0.0754